baseline/baseline.py

```diff
@@ -6,54 +6,76 @@
     seed = 0
+    epsilon = 1e-4
 
     def __init__(self):
         super().__init__()
         torch.random.manual_seed(self.seed)
-        self.mel = MelSpectrogram(
-            sample_rate=self.sample_rate, n_fft=self.n_fft, n_mels=self.n_mels
+
+        # Create a Hann window buffer to apply to frames prior to FFT.
+        self.register_buffer("window", torch.hann_window(self.n_fft))
+
+        # Create a mel filter buffer.
+        mel_scale = Tensor(
+            librosa.filters.mel(self.sample_rate, n_fft=self.n_fft, n_mels=self.n_mels)
         )
+        self.register_buffer("mel_scale", mel_scale)
+
+        # Projection matrices.
         self.emb4096 = torch.nn.Parameter(
             torch.rand(self.n_mels, 4096) / math.sqrt(self.n_mels)
         )
         self.emb2048 = torch.nn.Parameter(torch.rand(4096, 2048) / math.sqrt(4096))
         self.emb512 = torch.nn.Parameter(torch.rand(2048, 512) / math.sqrt(2048))
         self.emb128 = torch.nn.Parameter(torch.rand(512, 128) / math.sqrt(512))
         self.emb20 = torch.nn.Parameter(torch.rand(128, 20) / math.sqrt(128))
+
+        # An activation to squash the 20D embedding to a [0, 1] range.
         self.activation = torch.nn.Sigmoid()
 
-    def forward(self, x: Tensor, hop_size_samples: int, center: bool):
-        self.mel.hop_length = hop_size_samples
-        self.mel.center = center
-        self.mel.spectrogram.hop_length = hop_size_samples
-        self.mel.spectrogram.center = center
-        x = torch.log(self.mel(x) + 1e-4)
-        x = x.swapaxes(1, 2)
+    def forward(self, x: Tensor):
+        # Compute the real-valued Fourier transform on windowed input signal.
+        x = torch.fft.rfft(x * self.window)
+
+        # Convert to a power spectrum.
+        x = torch.abs(x) ** 2.0
+
+        # Apply the mel-scale filter to the power spectrum.
+        x = torch.matmul(x, self.mel_scale.transpose(0, 1))
+
+        # Convert to a log mel spectrum.
+        x = torch.log(x + self.epsilon)
+
+        # Apply projections to get all required embeddings
         x4096 = x.matmul(self.emb4096)
         x2048 = x4096.matmul(self.emb2048)
         x512 = x2048.matmul(self.emb512)
         x128 = x512.matmul(self.emb128)
         x20 = x128.matmul(self.emb20)
-        # [0, 1] range
+
+        # The 20-dimensional embedding is specified to be int8. To cast to int8 we'll
+        # apply an activation to ensure the embedding is in a 0 to 1 range first.
         x20 = self.activation(x20)
-        # Convert x20 to int8.
-        newmax = torch.iinfo(torch.int8).max
-        newmin = torch.iinfo(torch.int8).min
-        x20 = torch.tensor(
-            (x20 * (newmax - newmin) + newmin), dtype=torch.int8, device=x20.device
-        )
+
+        # Scale to int8 value range and cast to int
+        int8_max = torch.iinfo(torch.int8).max
+        int8_min = torch.iinfo(torch.int8).min
+        x20 = x20 * (int8_max - int8_min) + int8_min
+        x20 = x20.type(torch.int8)
```

Comment:
I was testing this with a some different input signals and I don't think this is working quite how we expect. The ranges of the embedded signals after projection seem to be getting pretty extreme (each projection seems to emphasize that difference even more), even though it looks like the projection matrices are being normalized. So by the time we get to x20 the values are pretty only -128 or 127, we are capturing the range in that embedding very well.

Reply:
I don't know if this is helpful, but you might just do individual projections from the largest embedding size down to each smaller one, instead of cascading projections. I suggested this a little while ago.

Reply:
Ya good idea -- i'll try this out